### PR TITLE
Add 'Thoughts of a Rustacean learning Go'

### PR DESCRIPTION
**Pull request by Manishearth**
This post gives an idea of what Go looks like from the point of view of someone coming from Rust. Whilst it's mainly intended for Rustaceans, it does try to give an insight into how things are done in Rust and Go to those familiar with neither.

Modified crosspost of http://inpursuitoflaziness.blogspot.in/2015/02/thoughts-of-rustacean-learning-go.html

I'm open to rewriting or expanding the post. Feel free to be brutal in the review :smile: 

cc @bcoe @markstory @aredridel @jameswomack comments appreciated!

### CODE REVIEW

**Comment by Manishearth** [comment]
Oh, and I haven't yet generated the site locally since I'm on a slow connection for the next few days (installing `npm` would be hard), so if someone could test out the rendering that would be great!

File: thoughts-of-a-rustacean-learning-go.md [hunk 1]
```diff
@@ -0,0 +1,244 @@
+<!--
+headline: Thoughts of a Rustacean learning Go
+description: the experiences of a Rustacean learning Go
+author: Manish Goregaokar
+datePublished: 2015-04-24
+twitter: ManishEarth
+github: manishearth
+-->
+
+// Might want to mention that this is a crosspost
```
Comment (bcoe) [inline thoughts-of-a-rustacean-learning-go.md]: I think that since we'll be making edits to the article for polyglotweekly.com, we probably don't need to explicitly mention that it's a cross post (unless you'd like to).

File: thoughts-of-a-rustacean-learning-go.md [hunk 2]
```diff
@@ -0,0 +1,244 @@
+<!--
+headline: Thoughts of a Rustacean learning Go
```
Comment (bcoe) [inline thoughts-of-a-rustacean-learning-go.md]: To give even more context, how do you feel about the title:

### A Rust Contributor Tries Their Hand at Go

Or something along those lines?

File: thoughts-of-a-rustacean-learning-go.md [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+and the [Servo browser engine][servo].
+
+// Perhaps a different form of introduction?
+
+
+[trpl]: http://rust-lang.org/
+[compiler]: https://github.com/rust-lang/rust
+[servo]: https://github.com/servo/servo
+
+Recently, for a course, we were told to use Go for the assignments and project. Whilst a tad disappointed
```
Comment (bcoe) [inline thoughts-of-a-rustacean-learning-go.md]: might rephrase to "we were asked".

File: thoughts-of-a-rustacean-learning-go.md [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+compared a lot as the "hot new languages", and finally I'd get to see the other side of the argument.
+
+By the end of the course, aside from some simpler assignments, I'd written
+[a basic implementation of the Raft consensus algorithm][raft] and a [toy application][dash] using it,
+both in Go.
+
+Before I get into the experience, let me preface this by mentioning that Rust and Go don't exactly target the same audiences.
+Go is garbage collected and is okay with losing out on some performance for ergonomics;
+whereas Rust tries to keep everything as a compile time check as much as possible.
+This usually makes Rust much more useful for lower level applications.
```
Comment (bcoe) [inline thoughts-of-a-rustacean-learning-go.md]: I might make this statement a bit less definitive, e.g., "This can make Rust better for lower level applications".

Is there something we could reference in this paragraph, perhaps:
- some performance benchmarks.
- or another comparative article, talking about the strengths of both languages?

File: thoughts-of-a-rustacean-learning-go.md [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+// I've structured this as "what I liked"/"what I didn't like", but I'm open to other ways of structuring this too
+
+## What I liked
+
+Despite the performance costs, having a GC at your disposal after using Rust for very long is quite liberating. 
+For a while my internalized borrow checker would throw red flags on me tossing around data indiscriminately,
+but I learned to  ignore it as far as Go code goes. I was able to quickly share state via pointers without
+worrying about safety, which was quite useful.
+
+Having channels as part of the language itself was also quite ergonomic.  The `data <- chan` and `chan <- data` syntax is fun to use,
```
Comment (bcoe) [inline thoughts-of-a-rustacean-learning-go.md]: It would be great to have a side-by side comparison of the two styles of syntax here, rather than the inline `data <- chan` and `.send()`.

File: thoughts-of-a-rustacean-learning-go.md [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+-->
+
+// Might want to mention that this is a crosspost
+
+
+For the past year or so I've been playing with the [Rust programming language][trpl].
+I've come to love the language and spend quite a bit of my free time contributing to the [compiler][compiler]
+and the [Servo browser engine][servo].
+
+// Perhaps a different form of introduction?
```
Comment (bcoe) [inline thoughts-of-a-rustacean-learning-go.md]: I like this introduction, it's neat to get some information about your background in Rust.

File: thoughts-of-a-rustacean-learning-go.md [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+and whilst it's not very different from `.send()` and `.recv()` in Rust, I found it surprisingly easy to read.
+Initially I got confused often by which side the channel was, but after a while I got used to it.
+It also has an in built select block for selecting over channels (Rust has a macro for the same purpose, but no language support).
+
+
+`gofmt`. The Go style of coding is different from the Rust one (tabs vs spaces, how declarations look),
+but I continued to use the Rust style because of the muscle memory (also too lazy to change the settings
+in my editor). `gofmt` made life easy since I could just run it in a directory and it would fix everything.
+Eventually I was able to learn the proper style by watching my code get corrected.
+I'd love to see a `rustfmt`, however!
```
Comment (bcoe) [inline thoughts-of-a-rustacean-learning-go.md:99]: I've recently started using the similar tool `standard` for my JavaScript libraries:

https://github.com/feross/standard

and love it.

File: thoughts-of-a-rustacean-learning-go.md [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+
+Go has really great <s>green threads</s> goroutines. They're rather efficient
+(I can spawn a thousand and it schedules them nicely), and easy to use.
+
+Go has really good in built support and tooling for tests (Rust does too).
+I enjoyed writing tests in Go quite a bit due to this.
+
+## What I didn't like
+
+Sadly, there are a lot of things here, but bear in mind what I mentioned above
```
Comment (bcoe) [inline thoughts-of-a-rustacean-learning-go.md]: I might rephrase this paragraph slightly, to make it less likely that you get negative feedback from Go enthusiasts, perhaps:

> There are quite a few things here, but bear in mind I'm new to Go, and am still learning the _Go way_ of doing things. I would love to hear feedback about how I can overcome some of the problems I ran into.

File: thoughts-of-a-rustacean-learning-go.md [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+
+This is extremely useful for sending typed messages across channels. For example, in Servo we use such
+[an enum for sending details about the progress of a fetch to the corresponding XHR object][servo-xhr-enum].
+[Another such enum][servo-constellation-enum] is used for communication between the constellation and the compositor/script.
+
+
+This gives us a great degree of type safety; I can send messages with different data within them,
+however I can only send messages that the other end will know how to handle since they must all be of the type of the message enum.
+
+In Go there's no obvious way to get this. The closest thing is the type called `interface {}`
```
Comment (bcoe) [inline thoughts-of-a-rustacean-learning-go.md:150]: I'd again be interested in seeing some larger side-by-side code comparisons here. It would help me get a better idea of how you would structure data for passing messages between threads in Rust vs. Go-routines in Go.

File: thoughts-of-a-rustacean-learning-go.md [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+but this will behave exactly like `interface{}` (implemented on all types) unless I add a dummy method to it, which seems hackish.
+This brings me to my next point:
+
+
+[servo-xhr-enum]: https://github.com/servo/servo/blob/d5dd1d658e5d79701fb9d028479a0fcb26a033fa/components/script/dom/xmlhttprequest.rs#L100
+[servo-constellation-enum]: https://github.com/servo/servo/blob/d5dd1d658e5d79701fb9d028479a0fcb26a033fa/components/msg/constellation_msg.rs#L198
+
+### Smart interfaces
+
+This is something many would consider a feature in Go, but from the point of view of a Rustacean, I'm rather annoyed by this.
```
Comment (bcoe) [inline thoughts-of-a-rustacean-learning-go.md]: annoyed has a fairly negative connotation. I might rephrase it to something like:

> coming from Rust, Smart interfaces felt almost too magical, and sometimes tripped me up during refactors.

File: thoughts-of-a-rustacean-learning-go.md [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+Having interfaces be autoimplemented on the basis of the method signature is a rather ergonomic feature in my opinion
+and it reduces boilerplate. However, it's just not what I'm used to and it restricts me from writing certain types of code without using
+dummy methods [like so][dummy] to get static type safety.
+
+
+[dummy]: https://github.com/Manishearth/cs733/blob/e86f22487f53b6bee21415d6b6ca56b8007792cc/assignment3/raft/raft.go#L57
+
+### Packages and imports
+
+Go puts severe restrictions on where I can put my files. All files in a folder are namespaced into the same package
```
Comment (bcoe) [inline thoughts-of-a-rustacean-learning-go.md:327]: I've heard a lot of critiques about how Go handles imports, and its lack of a package manager. Would love a follow up post at some point about how Rust's package manager/module system.

File: thoughts-of-a-rustacean-learning-go.md [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+however the behavior can be nearly arbitrarily overridden if necessary (you can even do [scary things like this][chtulhu-import]).
+
+[absolute-import]: https://github.com/Manishearth/cs733/blob/9842305bab2f8402dbd235c8512a05f57da1cc5f/assignment2/kvstore/server.go#L5
+[relative-import]: https://github.com/servo/servo/blob/c2fc6e311a0cb3ec4c702c77bb5d13f97bd19078/components/script/Cargo.toml#L19
+[chtulhu-import]: https://twitter.com/horse_rust/status/517408273750704128
+[rust-module]: http://doc.rust-lang.org/book/crates-and-modules.html
+
+### Documentation
+
+ Rust's libraries aren't yet well documented; but this is mostly because the libraries are
```
Comment (bcoe) [inline thoughts-of-a-rustacean-learning-go.md]: I think this paragraph sounds like it's making excuses. It's totally understandable that Rust's documentation needs work, given that the language's syntax has evolved significantly over the past couple years.

I might restructure it to something like:

> Full disclosure, Rust's libraries aren't yet well documented. This relates largely to the fact that many of the core libraries are still in flux. There are some great initiatives by developers like the awesome Steve Klabnik to improve the documentation.
> 
> Having said this, I also found that Go's documentation was skimpy in places.

File: thoughts-of-a-rustacean-learning-go.md [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+
+A major recurring point is that Go seems to advocate runtime over compile time checking (despite being a compiled language),
+something which is totally opposite to what Rust does. This is not just visible in language features
+(like the GC), but also in the tools that are provided — as mentioned above, Go does not give good tools for
+creating type safe abstractions, and the programmer must add dynamic matching over types to overcome this.
+This is similar (though not the same) as what languages like Python and Javascript advocate,
+however these are generally interpreted, not compiled (and come with the benefits of being interpreted),
+so there's a good tradeoff.
+
+Go isn't a language I intend to use for personal projects in the near future. I enjoyed it,
```
Comment (bcoe) [inline thoughts-of-a-rustacean-learning-go.md]: I think this conclusion might come across as a bit too combative, I might word it a bit more like:

> I wasn't an immediate convert to Go, having said this I really enjoyed learning, and playing with the language. I'm coming from a different paradigm and it would take me time to adjust. Having said this I highly recommend trying out the language.

**Comment by bcoe** [comment]
@Manishearth I really enjoyed this post, it was neat to hear the reaction of someone entrenched in one community giving another language a shot.

A couple comments:

### Tone Policing

I think we have to be particularly careful about using a non-combative tone anywhere in this type of article.

In a perfect world, I'd like to see someone from the Go community write a follow up post to your article. The last thing we want is it turning into an argument between the two communities.

I think you did a good job of keeping a positive tone, but I think we can be _even more paranoid_.

### Code Examples

You make a lot of great comparisons between the two languages, but I think more side-by-side code comparisons would really help tell the story... speaking of which:

I just added [Rust support](https://github.com/npm/marky-markdown/pull/44) to Polyglot Weekly \o/

File: thoughts-of-a-rustacean-learning-go.md (hunk 1 above)
Comment (Manishearth) [inline thoughts-of-a-rustacean-learning-go.md]: Don't mind either way.

File: thoughts-of-a-rustacean-learning-go.md (hunk 4 above)
Comment (Manishearth) [inline thoughts-of-a-rustacean-learning-go.md]: http://www.infoworld.com/article/2877924/application-development/go-rust-road-ahead-young-programming-languages.html says that Go isn't designed for systems programming, and we have some benchmarks at http://joshitech.blogspot.in/2014/11/go-vs-rust-productivity-vs-performance.html. Sound good?

File: thoughts-of-a-rustacean-learning-go.md (hunk 11 above)
Comment (Manishearth) [inline thoughts-of-a-rustacean-learning-go.md:327]: Noted. It's not too glamorous though, it just has a system of crates and modules (and http://doc.rust-lang.org/book/crates-and-modules.html probably explains it better)

File: thoughts-of-a-rustacean-learning-go.md (hunk 12 above)
Comment (Manishearth) [inline thoughts-of-a-rustacean-learning-go.md]: > It's totally understandable that Rust's documentation needs work, given that the language's syntax has evolved significantly over the past couple years.

Actually, that's not the issue, many of the libraries are new (in their current form, that is)and unstable. I'll reword it, but I'd like to keep the stability distinction.

File: thoughts-of-a-rustacean-learning-go.md (hunk 2 above)
Comment (Manishearth) [inline thoughts-of-a-rustacean-learning-go.md]: That sounds good. FWIW "Rustacean"=="Rust user" ("Gopher" == "Go user"), but the term is probably still obscure.

What should I use for the 'description'?

File: thoughts-of-a-rustacean-learning-go.md (hunk 7 above)
Comment (Manishearth) [inline thoughts-of-a-rustacean-learning-go.md:99]: Ooh, looks sweet :D

Any idea how I should work this into the text (if I should at all)?

File: thoughts-of-a-rustacean-learning-go.md (hunk 9 above)
Comment (Manishearth) [inline thoughts-of-a-rustacean-learning-go.md:150]: Done, though I focused on the message passing here -- didn't really go into thread/goroutine spawning. I can, though, if you want (just makes the code longer)

**Comment by Manishearth** [comment]
I addressed the individual comments (might add some more code examples later)

> Tone Policing

Yeah, I had kept this in mind whilst writing the original post too -- I tried to make sure things came off as "as a Rust user, X seems strange to me (though it's not a bad idea)" or whatever over "Go is silly because of X"

I'll try to stay paranoid :wink:

**Comment by Manishearth** [comment]
@dyzp might be interested in writing the flipside "Thoughts of a Gopher learning Rust" or whatever :smile: (see also: http://www.reddit.com/r/rust/comments/33ngl0/my_first_lockless_datastructure_in_rust/)

@nox is from the Erlang community and these days is poking at Servo, he might be able to write a similar post for Erlang-to-Rust

**Comment by Manishearth** [comment]
(Also, yay for the syntax highlighting support :grinning: )

File: thoughts-of-a-rustacean-learning-go.md [hunk 14]
```diff
@@ -0,0 +1,439 @@
...
+// On the sender end
+fn draw_thingy(tx: Sender<MyMessage) {
+    tx.send(DrawLine(Red, Point{x:0, y:0}, Point{x: 1, y: 1}));
+    tx.send(SetBackground(Blue))
+    // we can't send anything other than `MyMessage` variants    
+}
+
+fn main() {
+    // do some setup
+    let (rx, tx) = channel();
```
Comment (bcoe) [inline thoughts-of-a-rustacean-learning-go.md]: Just reading up on Rust as I edit, should this be:

`let (tx, rx)`?

https://doc.rust-lang.org/std/sync/mpsc/fn.channel.html

Comment (Manishearth) [inline thoughts-of-a-rustacean-learning-go.md]: Yep, fixed -- I'd added this section as an afterthought.

**Comment by bcoe** [comment]
@Manishearth I really enjoyed this second read through the article, the code examples helped pull things together for me (not knowing Go, or Rust it helped it gave me a starting point to start dissecting the syntax).

I figure give some other editors a chance to chime in, and then release the article Monday or Tuesday? I'm hoping we can keep up a momentum of one article a week.

**Comment by Manishearth** [comment]
Sounds good to me!

Oh, and you probably should learn Go if you have a lazy afternoon. Go is a language that can be picked up pretty easily (unlike Rust -- sure, you can pick up the basics easily, and the ownership/borrowing semantics take some time, but there are a lot of more complex features which means that you need some time to really get the language)

On that note, perhaps I should add a section to the article that mentions how easy Go is to pick up :)

**Comment by bcoe** [comment]
@Manishearth good advice, and good idea with regards to adding a small section.

I was just thinking while reading this article that I've been lazy about picking up languages recently.

**Comment by Manishearth** [comment]
added

**Comment by bcoe** [comment]
@Manishearth I like the additional section about how easy Go is to pick up :+1:

If you wouldn't mind deleting the two `//` comments in the post, I'll merge it when I get into work tomorrow? 9AM PDT.

Thanks for the great article.

**Comment by Manishearth** [comment]
Done!

**Comment by Manishearth** [comment]
I've been pushing minor fixes to master -- but is it fine if I open another PR with some slightly larger modifications which came up in the discussions? Or should we avoid changing a published post?

(I want to add a section on the compilation speed, and mention the idiomatic way of using interfaces. I can prefix both with an _Edit:_  if necessary)

**Comment by bcoe** [comment]
@Manishearth I have no objections to making some edits to the post :+1:, but if they're major let's do them as pull requests.

**Comment by savaki** [comment]
@Manishearth I really enjoyed the post as well.  I thought I might offer a type safe alternative that is a little more idiomatic go.

https://gist.github.com/savaki/c085cc86bb9dfb74c700
